### PR TITLE
EOS-25979: Deployment with init containers

### PR DESCRIPTION
**Pull request by cahuja123**
# Problem Statement
- CORTX N-Node Deployment with Init Containers

# Design
-  Convert mini provisioner into Init containers on the related deployments.
-  Rename the conatiners name w.r.t services framework.

# Coding
   Checklist for Author
-  [x] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
- [ ] Unit and System Tests are added
- [ ] Test Cases cover Happy Path, Non-Happy Path and Scalability
- [ ] Testing was performed with RPM

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
- [ ] Interface change (if any) are documented
- [ ] Side effects on other features (deployment/upgrade)
- [ ] Dependencies on other component(s)

# Review Checklist 
  Checklist for Author
- [x] JIRA number/GitHub Issue added to PR
- [x] PR is self reviewed
- [x] Jira and state/status is updated and JIRA is updated with PR link
- [x] Check if the description is clear and explained

# Documentation
  Checklist for Author
- [ ] Changes done to WIKI / Confluence page / Quick Start Guide

-----
[View rendered test/deploy/kubernetes/README.md](https://github.com/Seagate/cortx-prvsnr/blob/EOS-25979-Deployment_with_init_container/test/deploy/kubernetes/README.md)

### CODE REVIEW

**Comment by cortx-admin** [comment]
Can one of the admins verify this patch?

**Review by tanujashinde0405 — APPROVED** [review]

File: test/deploy/kubernetes/deploy.sh [hunk 1]
```diff
@@ -163,16 +201,20 @@ if [ $REDEFPODS = true ]; then
 
     # Create HA Node (POD)
     print_header "Creating HA Node - Cluster";
-    kubectl apply -f "$BASEPATH/provisioner-pods/deployment_ha_node.yaml" --namespace "$NAMESPACE";
+    kubectl apply -f "$BASEPATH/runtime-pods/components_ha_node.yaml" --namespace "$NAMESPACE";
 
     # Wait for HA Node (POD) execution
     printf "\nWaiting for HA Node.";
-    while [ "$(kubectl get pods | grep ha-node | awk '{print $3}')" != "Completed" ]; do
+    while [ "$(kubectl get pods | grep ha-node | awk '{print $3}')" != "Running" ]; do
```
Comment (nitin-seagate) [inline test/deploy/kubernetes/deploy.sh:208]: **Improvement:**
We should check for status such as `Failed` and `Unknown` as well

File: test/deploy/kubernetes/deploy.sh [hunk 2]
```diff
@@ -22,6 +22,31 @@ function show_usage {
     exit 1
 }
 
+function wait_for_pod_to_be_ready(){
```
Comment (nitin-seagate) [inline test/deploy/kubernetes/deploy.sh:25]: **Improvement:**
We should make function generic as in:
check_pod_status(message_header, filter_args, status_list):
message_header: will display message like "Check for pod `filter_args` with status `status_list`"
filter_args: Will be used to filter pods (Example: `data-node.-.*`)
status_list: List of status polling

File: test/deploy/kubernetes/deploy.sh [hunk 3]
```diff
@@ -99,6 +116,23 @@ kubectl create configmap solution-config \
     --from-file="$BASEPATH"/solution-config/config.yaml \
     --namespace "$NAMESPACE";
 
+# Create Config Map For Machine-id
+kubectl create configmap control-node-id --from-literal=id=9995f539f4f770e2a3fe9e2e615c32a8 --namespace "$NAMESPACE";
+if [ $REDEFPODS = true ]; then
+    kubectl create configmap ha-node-id --from-literal=id=1115f539f4f770e2a3fe9e2e615c32a8 --namespace "$NAMESPACE";
+    kubectl create configmap server-node1-id --from-literal=id=ddd120a9e051d103c164f605615c32a4 --namespace "$NAMESPACE";
```
Comment (nitin-seagate) [inline test/deploy/kubernetes/deploy.sh:123]: What's use of these, we nywaz have hard-coded values in pod definition files???

**Review by nitin-seagate — CHANGES_REQUESTED** [review]

File: test/deploy/kubernetes/deploy.sh (hunk 3 above)
Comment (cahuja123) [reply test/deploy/kubernetes/deploy.sh:123]: We have shifted from Pod to Deployment kind so we are not using downward API for placing machine id inside containers, Instead now we're using configmap (Same as like services team is using.)

File: test/deploy/kubernetes/README.md [hunk 4]
```diff
@@ -31,38 +31,33 @@ Run reimage.sh script to install helm, pull 3rd party as well as cortx-all docke
 
 Deploy Provisioner Deployment POD **(Run Only on Master Node)**  
 
-Run deploy.sh script to create cortx-deployment and Launch 1 cortx-provisioner deployment container on each POD. Once the script is executed, 3 Storage deployment POD’s will be running on 3 Nodes and 1 Control Pod will be running on master node. Each POD will have provisioner container which will execute cortx_setup config apply and cluster bootstrap commands.  
+Run deploy.sh script to create cortx-deployment and Launch 1 cortx-provisioner deployment container on each POD. Once the script is executed, 3 Storage deployment POD’s will be running on 3 Nodes and 1 Control Pod will be running on master node. Each POD will have 1 init provisioner container which will execute cortx_setup config apply and cluster bootstrap commands.
```
[the remark this answers is not in the document]
Comment (cahuja123) [reply test/deploy/kubernetes/README.md]: Done

File: test/deploy/kubernetes/deploy.sh [hunk 5]
```diff
@@ -99,6 +116,23 @@ kubectl create configmap solution-config \
     --from-file="$BASEPATH"/solution-config/config.yaml \
     --namespace "$NAMESPACE";
 
+# Create Config Map For Machine-id
+kubectl create configmap control-node-id --from-literal=id=9995f539f4f770e2a3fe9e2e615c32a8 --namespace "$NAMESPACE";
+if [ $REDEFPODS = true ]; then
+    kubectl create configmap ha-node-id --from-literal=id=1115f539f4f770e2a3fe9e2e615c32a8 --namespace "$NAMESPACE";
```
[the remark this answers is not in the document]
Comment (cahuja123) [reply test/deploy/kubernetes/deploy.sh:122]: Will take it up later.

File: test/deploy/kubernetes/deploy.sh [hunk 6]
```diff
@@ -146,11 +180,15 @@ kubectl apply -f "$BASEPATH/external-services/headless_control_node.yaml" --name
...
 # Create Control Node (POD)
 print_header "Creating Cotrol Node - Cluster";
-kubectl apply -f "$BASEPATH/provisioner-pods/deployment_control_node.yaml" --namespace "$NAMESPACE";
+kubectl apply -f "$BASEPATH/runtime-pods/components_control_node.yaml" --namespace "$NAMESPACE";
 
 # Wait for Control-Node (POD) execution
 printf "\nWaiting for Control Node.";
-while [ "$(kubectl get pods | grep control-node | awk '{print $3}')" != "Completed" ]; do
+while [ "$(kubectl get pods | grep control-node | awk '{print $3}')" != "Running" ]; do
+    if [ "$(kubectl get pods | grep control-node | awk '{print $3}')" == "Error" ]; then
```
[the remark this answers is not in the document]
Comment (cahuja123) [reply test/deploy/kubernetes/deploy.sh:188]: Will take it up later.

File: test/deploy/kubernetes/deploy.sh (hunk 1 above)
Comment (cahuja123) [reply test/deploy/kubernetes/deploy.sh:208]: Will take it up later.

File: test/deploy/kubernetes/deploy.sh (hunk 2 above)
Comment (cahuja123) [reply test/deploy/kubernetes/deploy.sh:25]: Will take it up later.

File: test/deploy/kubernetes/deploy.sh [hunk 7]
```diff
@@ -223,12 +271,15 @@ else
...
     done
+    printf "\nWaiting for Storage pod to be ready."
+    wait_for_pod_to_be_ready 'storage-node.-.*'
+    printf "\n\n"
 fi
 
 echo -e "Monitor All Nodes for Completed status";
 kubectl get pods -o wide -n "$NAMESPACE";
-exit 0;
+exit 0;
```
[the remark this answers is not in the document]
Comment (cahuja123) [reply test/deploy/kubernetes/deploy.sh]: Done

**Review by nitin-seagate — APPROVED** [review]
LGTM

**Pull request merged**